mu4/languages/internal/languagescontroller.cpp

```diff
@@ -0,0 +1,374 @@
+{
+    RetVal<QString> result;
+
+    QString languageArchivePath = configuration()->languageArchivePath(languageCode);
+
+    QBuffer buff;
+    INetworkManagerPtr networkManagerPtr = networkManagerCreator()->makeNetworkManager();
+
+    async::Channel<Progress> downloadChannel = networkManagerPtr->downloadProgressChannel();
+    downloadChannel.onReceive(new deto::async::Asyncable(), [&progressChannel](const Progress& progress) {
+        progressChannel.send(LanguageProgress(DOWNLOADING_STATUS, progress.current,
+                                              progress.total));
+    });
+
+    Ret getLanguage = networkManagerPtr->get(configuration()->languageFileServerUrl(languageCode), &buff);
+    if (!getLanguage) {
+        LOGE() << "Error save file";
+        result.ret = make_ret(Err::ErrorDownloadLanguage);
+        return result;
+    }
+
+    QFile file(languageArchivePath);
+    file.open(QIODevice::WriteOnly);
+    file.write(buff.data());
+    file.close();
+
+    result.ret = make_ret(Err::NoError);
+    result.val = languageArchivePath;
+    return result;
+}
+
+Ret LanguagesController::removeLanguage(const QString& languageCode) const
+{
+    QStringList files = configuration()->languageFilePaths(languageCode);
+
+    for (const QString& filePath: files) {
+        Ret ret = fsOperations()->remove(filePath);
+        if (!ret) {
+            LOGE() << "Error remove file" << filePath << ret.code() << ret.text();
+            return make_ret(Err::ErrorRemoveLanguageDirectory);
+        }
+    }
+
+    return make_ret(Err::NoError);
+}
+
+Ret LanguagesController::loadLanguage(const QString& languageCode)
+{
+    QStringList files = configuration()->languageFilePaths(languageCode);
+
+    for (const QString& filePath: files) {
+        QTranslator* translator = new QTranslator;
+        bool ok = translator->load(filePath);
+        if (ok) {
+            qApp->installTranslator(translator);
+            m_translatorList.append(translator);
+        } else {
+            LOGE() << "Error load translate" << filePath;
+            delete translator;
+        }
+    }
+
+    QLocale locale(languageCode);
+    QLocale::setDefault(locale);
+    qApp->setLayoutDirection(locale.textDirection());
+
+    return make_ret(Err::NoError);
+}
+
+void LanguagesController::resetLanguageByDefault()
+{
+    Ret load = loadLanguage(DEFAULT_LANGUAGE);
+    if (!load) {
+        return;
+    }
+
+    configuration()->setCurrentLanguageCode(DEFAULT_LANGUAGE);
+}
+
+void LanguagesController::th_install(const QString& languageCode, async::Channel<LanguageProgress> progressChannel,
+                                     std::function<void(const QString&, const Ret&)> callback)
+{
+    progressChannel.send(LanguageProgress(ANALYSING_STATUS, true));
+
+    RetVal<QString> download = downloadLanguage(languageCode, progressChannel);
+    if (!download.ret) {
+        callback(languageCode, download.ret);
```

Comment:
Here may be a problem.
Callback will be called in this thread, and not class thread (where set callback), implementation of callback maybe don't expect this (no mutex etc). 
I suggestion use channel for notification about finished and result.

Reply:
fixed